### PR TITLE
[SE-0031] Adjusting 'inout' Declarations for Type Decoration

### PR DESCRIPTION
**Pull request by dduan**
As the title indicates, this PR implements [SE-0031](https://github.com/apple/swift-evolution/blob/master/proposals/0031-adjusting-inout-declarations.md).
- [x] A warning is added for old `inout` position.
- [x] fixit for removing old inout position and inserting it in new position specified in `SE-0031`
- [x] when `inout` appears in both position, only the fixit for removal is issued
- [x] various places where inout declaration is printed
- [x] all tests are updated and passing.

### CODE REVIEW

File: lib/Parse/ParsePattern.cpp [hunk 1]
```diff
@@ -186,9 +186,10 @@ Parser::parseParameterClause(SourceLoc &leftParenLoc,
         status |= makeParserCodeCompletionStatus();
       }
     }
-
+    SourceLoc depercatedInoutLoc;
```
Comment (chriseidhof) [inline lib/Parse/ParsePattern.cpp]: There's a typo here: should be `deprecatedInoutLoc`

Comment (dduan) [inline lib/Parse/ParsePattern.cpp]: @chriseidhof fixed. Thanks Chris!

**Comment by lattner** [comment]
Totally awesome, thank you for working on this @dduan.  I have some specific comments which I'll comment on the diff itself.

File: include/swift/AST/DiagnosticsParse.def [hunk 2]
```diff
@@ -145,6 +145,9 @@ WARNING(lex_editor_placeholder_in_playground,none,
 
 NOTE(note_in_decl_extension,none,
      "in %select{declaration|extension}0 of %1", (bool, Identifier))
+WARNING(inout_as_attr_deprecated,none,
+        "'inout' as parameter attribute is deprecated, prefix type with it instead.",
```
Comment (lattner) [inline include/swift/AST/DiagnosticsParse.def]: This reads awkwardly, and diagnostic messages should not end with a period.  I'd suggest something like:

'inout' before a parameter name is deprecated, place it before the parameter type instead

File: lib/AST/ASTPrinter.cpp [hunk 3]
```diff
@@ -718,20 +720,18 @@ void PrintAST::printTypedPattern(const TypedPattern *TP) {
       Printer << "inout ";
     }
```
Comment (lattner) [inline lib/AST/ASTPrinter.cpp]: I think you can now remove  the special case "if (auto *IOT =" entirely.

File: lib/AST/ASTPrinter.cpp [hunk 4]
```diff
@@ -718,20 +720,18 @@ void PrintAST::printTypedPattern(const TypedPattern *TP) {
...
   }
 
+
+  printPattern(TP->getSubPattern());
+  Printer << ": ";
   Type T = TP->getType();
   if (auto *IOT = T->getAs<InOutType>()) {
     T = IOT->getObjectType();
     Printer << "inout ";
   }
```
Comment (lattner) [inline lib/AST/ASTPrinter.cpp]: Likewise, I think this "if" special case can be removed as well.

File: lib/AST/ASTPrinter.cpp [hunk 5]
```diff
@@ -1865,22 +1881,6 @@ void PrintAST::printOneParameter(const ParamDecl *param, bool Curried,
     }
   }
```
Comment (lattner) [inline lib/AST/ASTPrinter.cpp]: Similarly, both the InOutTypeRepr & the InOutType sides of this should be able to be dissolved now.  This probably allows this whole blob of code to go away (always a good sign :-)

File: lib/AST/ASTPrinter.cpp [hunk 6]
```diff
@@ -2831,16 +2831,16 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
...
       if (TD.hasName()) {
         Printer.printName(TD.getName(), PrintNameContext::FunctionParameter);
         Printer << ": ";
       }
-
+      if (auto *IOT = EltType->getAs<InOutType>()) {
+        Printer << "inout ";
+        EltType = IOT->getObjectType();
+      }
       if (TD.isVararg()) {
```
Comment (lattner) [inline lib/AST/ASTPrinter.cpp]: Another one, the special case for inout should hopefully be possible to drop entirely.

File: lib/Parse/ParsePattern.cpp [hunk 7]
```diff
@@ -186,9 +186,10 @@ Parser::parseParameterClause(SourceLoc &leftParenLoc,
...
       }
     }
-
+    SourceLoc deprecatedInoutLoc;
     // ('inout' | 'let' | 'var')?
     if (Tok.is(tok::kw_inout)) {
+      deprecatedInoutLoc = Tok.getLoc();
       param.LetVarInOutLoc = consumeToken();
       param.SpecifierKind = ParsedParameter::InOut;
     } else if (Tok.is(tok::kw_let)) {
```
Comment (lattner) [inline lib/Parse/ParsePattern.cpp]: I don't think you need the deprecatedInoutLoc variable at all here.  In the cases that use it, you can just use param.LetVarInOutLoc.  In the test for "deprecatedInoutLoc.isValid()", you can just check for "param.SpecifierKind == ParsedParameter::InOut".

**Comment by lattner** [comment]
That's all I see.  The test suite changes all look great, please send a patch to update the standard library as well.  Thank you again for implementing this @dduan!

File: lib/AST/ASTPrinter.cpp (hunk 5 above)
Comment (dduan) [inline lib/AST/ASTPrinter.cpp]: This one is different from the others. Skipping either branch causes a bunch of tests to fail.

**Comment by dduan** [comment]
Thanks for the review, Chris! I've addressed your comments to my best extent and pushed the changes up.

I have the patch for stdlib ready to go. Just waiting for this one to merge :)

**Comment by lattner** [comment]
Awesome, thanks again!

**Pull request merged**